File: backend/src/services/parsers/nonDepositSms.ts
/**
 * Filter provider SMS that are not deposit credits (PIN/OTP/greeting).
 * BlueVoucher/PayPulse often send a second "Dear customer…" PIN SMS after the credit.
 */

export function isNonDepositSms(raw: string): boolean {
  const t = raw.replace(/\s+/g, " ").trim();
  if (!t) return true;

  const looksLikeCredit =
    /credited\s+with|is credited|sent you|received\s+(an?\s+)?(easy\s*wallet|nad|n[$¤]|ewallet)|a\/c\s+\d+/i.test(
      t
    );

  // Explicit BlueVoucher cash-out PIN SMS (never a deposit credit)
  if (/bluevoucher\s+pin\b/i.test(t) && !looksLikeCredit) {
    return true;
  }

  // Greeting / PIN follow-up without a credit line
  if (/^dear\s+customer\b/i.test(t) && !looksLikeCredit) {
    return true;
  }

  // Standalone OTP / PIN messages (not FNB "sent you … PIN 12345" cash-out credits)
  if (/\b(otp|one[-\s]?time\s+pin|your\s+pin)\b/i.test(t) && !looksLikeCredit) {
    return true;
  }

  if (/\bpin\b/i.test(t) && /valid for\s+\d+/i.test(t) && !looksLikeCredit) {
    return true;
  }

  return false;
}

File: android/app/src/main/java/com/ewallet/capture/util/NonDepositSms.kt
package com.ewallet.capture.util

/**
 * Drop provider follow-up SMS that are not deposit credits
 * (BlueVoucher "Dear customer" PIN, standalone OTP, etc.).
 */
object NonDepositSms {
    fun shouldIgnore(raw: String): Boolean {
        val t = raw.replace(Regex("\\s+"), " ").trim()
        if (t.isEmpty()) return true

        val looksLikeCredit = Regex(
            """credited\s+with|is credited|sent you|received\s+(an?\s+)?(easy\s*wallet|nad|n[$¤]|ewallet)|a/c\s+\d+""",
            RegexOption.IGNORE_CASE
        ).containsMatchIn(t)

        // Explicit BlueVoucher cash-out PIN SMS (never a deposit credit)
        if (Regex("""bluevoucher\s+pin\b""", RegexOption.IGNORE_CASE).containsMatchIn(t) && !looksLikeCredit) {
            return true
        }

        if (Regex("""^dear\s+customer\b""", RegexOption.IGNORE_CASE).containsMatchIn(t) && !looksLikeCredit) {
            return true
        }
        if (Regex("""\b(otp|one[-\s]?time\s+pin|your\s+pin)\b""", RegexOption.IGNORE_CASE).containsMatchIn(t) &&
            !looksLikeCredit
        ) {
            return true
        }
        if (Regex("""\bpin\b""", RegexOption.IGNORE_CASE).containsMatchIn(t) &&
            Regex("""valid for\s+\d+""", RegexOption.IGNORE_CASE).containsMatchIn(t) &&
            !looksLikeCredit
        ) {
            return true
        }
        return false
    }
}
